### PR TITLE
Implement export of NVM files

### PR DESCRIPTION
**Pull request by magcks**

### CODE REVIEW

File: libs/mve/bundle_io.cc [hunk 1]
```diff
@@ -61,6 +62,14 @@ namespace
         rot[8] = 1.0f - 2.0f * q[1] * q[1] - 2.0f * q[2] * q[2];
         return rot;
     }
+
+    void get_quaternion_from_rot(const math::Matrix3d &values, double *rot)
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: `math::Matrix3d const& values`
Also, please just return a vector.

File: libs/mve/bundle_io.cc [hunk 2]
```diff
@@ -61,6 +62,14 @@ namespace
...
+    {
+        double w = rot[0] = std::sqrt(1.0 + values(0, 0) + values(1, 1) + values(2, 2)) / 2.0;
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: This is ugly. Remove `w` as it's not needed.

File: libs/mve/bundle_io.cc [hunk 3]
```diff
@@ -480,4 +489,77 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
     out.close();
 }
 
+void
+save_nvm_bundle (Bundle::ConstPtr bundle, const std::vector<NVMCameraInfo> &camera_info, std::string const& filename)
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: fix type, max 80 chars per line

File: libs/mve/bundle_io.cc [hunk 4]
```diff
@@ -480,4 +489,77 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
...
+
+    std::cout << "Writing bundle (" << cameras.size() << " cameras, "
+        << features.size() << " features): " << filename << "...\n";
+
+    std::ofstream out(filename.c_str(), std::ios::binary);
+    if (!out.good())
+        throw util::FileException(filename, std::strerror(errno));
+
+    out.precision(15);
+
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: Remove empty line

File: libs/mve/bundle_io.cc [hunk 5]
```diff
@@ -480,4 +489,77 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
...
+    out << "NVM_V3 \n\n";
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: Why the space before the newline?

File: libs/mve/bundle_io.cc [hunk 6]
```diff
@@ -480,4 +489,77 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
...
+        if (cam.flen == 0.0f || !camera_valid)  {
+            out << "0 0 0 0 0 0 0 0 0 0\n";
+            continue;
+        }
+
+        double rotq[4];
+        math::Matrix3d rot(math::Matrix3f(cam.rot));
+        get_quaternion_from_rot(rot, rotq);
+        math::Matrix3d rotinv = math::matrix_inverse(rot);
+        math::Vec3d center = -(rotinv * math::Vec3f(cam.trans[0], cam.trans[1], cam.trans[2]));
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: `math::Vec3f(cam.trans)`

File: libs/mve/bundle_io.cc [hunk 7]
```diff
@@ -480,4 +489,77 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
...
+
+        out << cam.flen << ' ' << rotq[0] << ' ' << rotq[1] << ' ' << rotq[2] << ' ' << rotq[3] << ' ' << center[0] << ' ' << center[1] << ' ' << center[2] << ' ' << cam.dist[0] << " 0\n";
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: 80 chars max

File: libs/mve/bundle_io.cc [hunk 8]
```diff
@@ -480,4 +489,77 @@ save_photosynther_bundle (Bundle::ConstPtr bundle, std::string const& filename)
...
+    }
+
+    out << '\n' << features.size() << '\n';
+
+    /* Write all features to bundle file. */
+    for (std::size_t i = 0; i < features.size(); ++i)
+    {
+        const Bundle::Feature3D &feature = features[i];
+        for (int j = 0; j < 3; ++j) {
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: Format bracket, or remove them

File: libs/mve/bundle_io.h [hunk 9]
```diff
@@ -80,6 +80,13 @@ void
 save_photosynther_bundle (Bundle::ConstPtr bundle,
     std::string const& filename);
 
+
+/**
+ * Writes a NVM bundle file.
+ */
+void
+save_nvm_bundle (Bundle::ConstPtr bundle, const std::vector<NVMCameraInfo> &camera_info, std::string const& filename);
```
Comment (simonfuhrmann) [inline libs/mve/bundle_io.h]: 80 chars

File: libs/mve/bundle_io.cc (hunk 3 above)
Comment (simonfuhrmann) [inline libs/mve/bundle_io.cc]: Passing a barely useful `camera_info` is ugly. Isn't there a better solution?

**Comment by simonfuhrmann** [comment]
Not a huge fan of this PR. Where is this required? It seems it's just adding code that is nowhere used.

**Comment by magcks** [comment]
it's used to export bundles to visualsfm. cmpmvs only accepts visualsfm bundels as input.

File: libs/mve/bundle_io.cc (hunk 3 above)
Comment (magcks) [inline libs/mve/bundle_io.cc]: `Bundle::Ptr
load_nvm_bundle (std::string const& filename,
    std::vector<NVMCameraInfo>* camera_info = nullptr);`

Passing a barely useful CameraInfo is ugly here as well. It's obvious, that I need all the data you created during the import for the export.

**Comment by simonfuhrmann** [comment]
Any updates to the comments?

**Comment by simonfuhrmann** [comment]
Any updates?

**Comment by simonfuhrmann** [comment]
Closing due to inactivity.

**Pull request closed without merging**